File: frontend/src/types/task.ts
// Task types matching backend API

export type TaskPriority = 'high' | 'medium' | 'low';

export type TaskCategory = 'bills' | 'savings' | 'review' | 'investment' | 'budget' | 'other';

export type RecurringFrequency = 'daily' | 'weekly' | 'monthly';

export interface Task {
  id: string;
  user_id: string;
  title: string;
  description?: string | null;
  priority: TaskPriority;
  category: TaskCategory;
  due_date?: string | null;
  is_recurring: boolean;
  recurring_frequency?: RecurringFrequency | null;
  is_completed: boolean;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority?: TaskPriority;
  category?: TaskCategory;
  due_date?: string;
  is_recurring?: boolean;
  recurring_frequency?: RecurringFrequency;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: TaskPriority;
  category?: TaskCategory;
  due_date?: string;
  is_recurring?: boolean;
  recurring_frequency?: RecurringFrequency;
  is_completed?: boolean;
}

export interface TaskSummary {
  total_tasks: number;
  active_tasks: number;
  completed_tasks: number;
  overdue_count: number;
  due_soon_count: number;
  overdue_tasks: Array<{
    id: string;
    title: string;
    due_date: string | null;
    priority: TaskPriority;
  }>;
  due_soon_tasks: Array<{
    id: string;
    title: string;
    due_date: string | null;
    priority: TaskPriority;
  }>;
}

export interface TaskFilters {
  completed?: boolean;
  priority?: TaskPriority;
  category?: TaskCategory;
  skip?: number;
  limit?: number;
}

File: backend/src/models/task.py
"""Task model for financial tasks and reminders."""

from datetime import date, datetime
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4

from sqlmodel import Field, SQLModel


class TaskPriority(str, Enum):
    """Task priority enumeration."""

    high = "high"
    medium = "medium"
    low = "low"


class TaskCategory(str, Enum):
    """Task category enumeration."""

    bills = "bills"
    savings = "savings"
    review = "review"
    investment = "investment"
    budget = "budget"
    other = "other"


class RecurringFrequency(str, Enum):
    """Recurring frequency enumeration."""

    daily = "daily"
    weekly = "weekly"
    monthly = "monthly"


class TaskBase(SQLModel):
    """Base task schema with common fields."""

    title: str = Field(max_length=200)
    description: Optional[str] = Field(default=None, max_length=1000)
    priority: TaskPriority = Field(default=TaskPriority.medium)
    category: TaskCategory = Field(default=TaskCategory.other)
    due_date: Optional[date] = None
    is_recurring: bool = Field(default=False)
    recurring_frequency: Optional[RecurringFrequency] = None


class Task(TaskBase, table=True):
    """Task database model."""

    __tablename__ = "tasks"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    user_id: UUID = Field(foreign_key="users.id", index=True)
    is_completed: bool = Field(default=False, index=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: Optional[datetime] = None


class TaskCreate(TaskBase):
    """Task creation schema."""

    pass


class TaskRead(TaskBase):
    """Task response schema."""

    id: UUID
    user_id: UUID
    is_completed: bool
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime]


class TaskUpdate(SQLModel):
    """Task update schema."""

    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[TaskPriority] = None
    category: Optional[TaskCategory] = None
    due_date: Optional[date] = None
    is_recurring: Optional[bool] = None
    recurring_frequency: Optional[RecurringFrequency] = None
    is_completed: Optional[bool] = None
